File: pages/reports.py
import streamlit as st
from datetime import datetime
from database import get_all_trades, get_setting
from calculations import compute_metrics
from exports import to_csv, to_excel, to_pdf

sar_rate = float(get_setting("usd_sar_rate") or 3.75)

st.markdown('<div class="page-header">Reports & Export</div>', unsafe_allow_html=True)

# Filters
with st.expander("Filter Dataset", expanded=True):
    fc1, fc2, fc3, fc4 = st.columns(4)
    with fc1:
        f_ticker = st.text_input("Ticker", key="rep_ticker")
        f_direction = st.selectbox("Direction", ["All", "Long", "Short"], key="rep_dir")
    with fc2:
        f_type = st.selectbox("Trade Type", ["All", "Day Trade", "Swing Trade", "Position Trade", "Long Hold"], key="rep_type")
        f_outcome = st.selectbox("Outcome", ["All", "Win", "Loss"], key="rep_outcome")
    with fc3:
        f_date_from = st.date_input("From Date", value=None, key="rep_from")
    with fc4:
        f_date_to = st.date_input("To Date", value=None, key="rep_to")

filters = {}
if f_ticker:
    filters["ticker"] = f_ticker
if f_direction != "All":
    filters["direction"] = f_direction
if f_type != "All":
    filters["trade_type"] = f_type
if f_outcome != "All":
    filters["outcome"] = f_outcome
if f_date_from:
    filters["date_from"] = f_date_from.isoformat()
if f_date_to:
    filters["date_to"] = f_date_to.isoformat()

trades = get_all_trades(filters)
metrics = compute_metrics(trades)
timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")

st.info(f"Dataset: **{len(trades)} trades** | Total P&L: **${metrics['total_pnl']:,.2f}** | Win Rate: **{metrics['win_rate']:.1f}%**")

if not trades:
    st.warning("No trades in the current filter. Adjust filters to include data.")
    st.stop()

st.markdown("---")
st.subheader("Download Reports")

col1, col2, col3 = st.columns(3)

with col1:
    st.markdown("**CSV Export**")
    st.caption("Raw trade data as comma-separated values.")
    csv_bytes = to_csv(trades, sar_rate)
    st.download_button(
        "Download CSV", csv_bytes,
        file_name=f"trades_{timestamp}.csv",
        mime="text/csv",
        use_container_width=True,
    )

with col2:
    st.markdown("**Excel Export**")
    st.caption("Multi-sheet workbook: Summary + Trade History + Monthly.")
    excel_bytes = to_excel(trades, sar_rate, metrics)
    st.download_button(
        "Download Excel", excel_bytes,
        file_name=f"trades_{timestamp}.xlsx",
        mime="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        use_container_width=True,
    )

with col3:
    st.markdown("**PDF Report**")
    st.caption("Formatted report with summary metrics and trade history.")
    pdf_bytes = to_pdf(trades, sar_rate, metrics)
    st.download_button(
        "Download PDF", pdf_bytes,
        file_name=f"trades_{timestamp}.pdf",
        mime="application/pdf",
        use_container_width=True,
    )

st.markdown("---")

# Summary preview
st.subheader("Report Preview — Performance Summary")
m = metrics
pf = m["profit_factor"]

sc1, sc2 = st.columns(2)
with sc1:
    rows = [
        ("Total Trades", str(m["total_trades"])),
        ("Winning Trades", str(m["winning_trades"])),
        ("Losing Trades", str(m["losing_trades"])),
        ("Win Rate", f"{m['win_rate']:.1f}%"),
        ("Profit Factor", "∞" if pf is None else f"{pf:.2f}"),
    ]
    for label, val in rows:
        c_l, c_v = st.columns([2, 1])
        c_l.markdown(f"**{label}**")
        c_v.markdown(val)

with sc2:
    rows2 = [
        ("Total P&L (USD)", f"${m['total_pnl']:,.2f}"),
        ("Total P&L (SAR)", f"SAR {m['total_pnl'] * sar_rate:,.2f}"),
        ("Avg Winner (USD)", f"${m['avg_winner']:,.2f}"),
        ("Avg Loser (USD)", f"-${m['avg_loser']:,.2f}"),
        ("Expectancy (USD)", f"${m['expectancy']:,.2f}"),
    ]
    for label, val in rows2:
        c_l, c_v = st.columns([2, 1])
        c_l.markdown(f"**{label}**")
        c_v.markdown(val)

File: static/js/reports.js
let repFilters = {};

registerPage('reports', {
  render() {
    return `
<div class="page-header">Reports &amp; Export</div>

<div class="filter-bar mb-6">
  <div class="filter-grid">
    <div class="form-group">
      <label>Ticker</label>
      <input type="text" id="rep-ticker" placeholder="AAPL" value="${repFilters.ticker || ''}">
    </div>
    <div class="form-group">
      <label>Direction</label>
      <select id="rep-direction">
        <option value="">All</option>
        <option value="Long" ${repFilters.direction === 'Long' ? 'selected' : ''}>Long</option>
        <option value="Short" ${repFilters.direction === 'Short' ? 'selected' : ''}>Short</option>
      </select>
    </div>
    <div class="form-group">
      <label>Trade Type</label>
      <select id="rep-type">
        <option value="">All</option>
        ${['Day Trade','Swing Trade','Position Trade','Long Hold'].map(t =>
          `<option value="${t}" ${repFilters.trade_type === t ? 'selected' : ''}>${t}</option>`).join('')}
      </select>
    </div>
    <div class="form-group">
      <label>Outcome</label>
      <select id="rep-outcome">
        <option value="">All</option>
        <option value="Win" ${repFilters.outcome === 'Win' ? 'selected' : ''}>Win</option>
        <option value="Loss" ${repFilters.outcome === 'Loss' ? 'selected' : ''}>Loss</option>
      </select>
    </div>
    <div class="form-group">
      <label>From</label>
      <input type="date" id="rep-from" value="${repFilters.date_from || ''}">
    </div>
    <div class="form-group">
      <label>To</label>
      <input type="date" id="rep-to" value="${repFilters.date_to || ''}">
    </div>
    <div class="form-group" style="align-self:end">
      <button class="btn btn-primary btn-full" id="rep-apply">Apply Filters</button>
    </div>
  </div>
</div>

<div id="rep-content"><div class="loading">Loading…</div></div>`;
  },

  async init() {
    await loadReports();

    document.getElementById('rep-apply').addEventListener('click', async () => {
      repFilters = {
        ticker:     document.getElementById('rep-ticker').value.trim(),
        direction:  document.getElementById('rep-direction').value,
        trade_type: document.getElementById('rep-type').value,
        outcome:    document.getElementById('rep-outcome').value,
        date_from:  document.getElementById('rep-from').value,
        date_to:    document.getElementById('rep-to').value,
      };
      await loadReports();
    });
  },
});

async function loadReports() {
  const q = buildQuery(repFilters);
  const [trades, m] = await Promise.all([
    API.get('/api/trades' + q),
    API.get('/api/metrics' + q),
  ]);

  const div = document.getElementById('rep-content');
  const pf = Fmt.inf(m.profit_factor);
  const cnt = trades.length;

  div.innerHTML = `
<div class="alert" style="background:var(--surface2);border:1px solid var(--border);color:var(--text-dim);margin-bottom:20px">
  Dataset: <strong>${cnt} trade${cnt !== 1 ? 's' : ''}</strong> ·
  Total P&amp;L: <strong class="${Fmt.cls(m.total_pnl)}">${Fmt.usd(m.total_pnl, false)}</strong> ·
  Win Rate: <strong>${Fmt.pct1(m.win_rate)}</strong>
  ${!cnt ? ' — <span class="text-red">No trades match current filters.</span>' : ''}
</div>

<div class="grid-3 mb-6">
  <div class="card" style="text-align:center">
    <div class="card-title">CSV Export</div>
    <p style="font-size:13px;color:var(--text-dim);margin-bottom:16px">Raw trade data as comma-separated values. Open in Excel, Google Sheets, or any spreadsheet app.</p>
    <button class="btn btn-primary btn-full" id="dl-csv" ${!cnt ? 'disabled' : ''}>Download CSV</button>
  </div>
  <div class="card" style="text-align:center">
    <div class="card-title">Excel Export</div>
    <p style="font-size:13px;color:var(--text-dim);margin-bottom:16px">Multi-sheet workbook with Summary, Trade History, and Monthly Performance sheets.</p>
    <button class="btn btn-primary btn-full" id="dl-excel" ${!cnt ? 'disabled' : ''}>Download Excel</button>
  </div>
  <div class="card" style="text-align:center">
    <div class="card-title">PDF Report</div>
    <p style="font-size:13px;color:var(--text-dim);margin-bottom:16px">Formatted report with performance summary and full trade history table.</p>
    <button class="btn btn-primary btn-full" id="dl-pdf" ${!cnt ? 'disabled' : ''}>Download PDF</button>
  </div>
</div>

<div class="card">
  <div class="card-title">Performance Summary Preview</div>
  <div class="grid-2">
    <div>
      ${[
        ['Total Trades',     cnt],
        ['Winning Trades',   m.winning_trades],
        ['Losing Trades',    m.losing_trades],
        ['Win Rate',         Fmt.pct1(m.win_rate)],
        ['Profit Factor',    pf],
        ['Largest Winner',   Fmt.usd(m.largest_winner, false)],
        ['Largest Loser',    Fmt.usd(m.largest_loser, false)],
      ].map(([l,v]) => `<div class="detail-row"><span class="detail-label">${l}</span><span class="detail-value">${v}</span></div>`).join('')}
    </div>
    <div>
      ${[
        ['Total P&L (USD)',   Fmt.usd(m.total_pnl, false)],
        ['Total P&L (SAR)',   Fmt.sar(m.total_pnl, sarRate)],
        ['Avg Winner (USD)',  Fmt.usd(m.avg_winner, false)],
        ['Avg Loser (USD)',   '-' + Fmt.usd(m.avg_loser, false)],
        ['Expectancy (USD)',  Fmt.usd(m.expectancy, false)],
        ['Gross Profit',      Fmt.usd(m.gross_profit, false)],
        ['Gross Loss',        Fmt.usd(m.gross_loss, false)],
      ].map(([l,v]) => `<div class="detail-row"><span class="detail-label">${l}</span><span class="detail-value">${v}</span></div>`).join('')}
    </div>
  </div>
</div>`;

  if (cnt) {
    document.getElementById('dl-csv').addEventListener('click',   () => { window.location.href = '/api/export/csv'   + q; });
    document.getElementById('dl-excel').addEventListener('click', () => { window.location.href = '/api/export/excel' + q; });
    document.getElementById('dl-pdf').addEventListener('click',   () => { window.location.href = '/api/export/pdf'   + q; });
  }
}
